### ISSUE TITLE
Add STS endpoint AssumeRoleWithSAML (assume_role_with_saml)

### ISSUE THREAD

**Issue by microe** [spulec/moto#2903]
I have a module that wraps our federation service to get a credential through STS using assume_role_with_saml. I needed to test this code so I took a stab at adding this endpoint. I have a solution for this in PR #2894, but the CI is not passing all tests.

I am able to get the tests to pass on both Python 2.7 and 3.8 when I call make test-only within my own tree, so I am a little stumped as to what else I need to do.

Here is a small example of the 2.7 build not working:
```
test_sts.test_assume_role_with_saml ... 172.17.0.1 - - [16/Apr/2020 19:14:48] "POST /moto-api/reset HTTP/1.1" 200 -

172.17.0.1 - - [16/Apr/2020 19:14:48] "POST / HTTP/1.1" 200 -

ERROR
***way farther down the log***
=====================================================================

ERROR: test_sts.test_assume_role_with_saml

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/virtualenv/python2.7.17/lib/python2.7/site-packages/nose/case.py", line 197, in runTest

    self.test(*self.arg)

  File "/home/travis/virtualenv/python2.7.17/lib/python2.7/site-packages/freezegun/api.py", line 735, in wrapper

    result = func(*args, **kwargs)

  File "/home/travis/build/spulec/moto/moto/core/models.py", line 95, in wrapper

    result = func(*args, **kwargs)

  File "/home/travis/build/spulec/moto/tests/test_sts/test_sts.py", line 203, in test_assume_role_with_saml

    SAMLAssertion=b64encode(saml_assertion.encode("utf-8")).decode("utf-8"),

  File "/home/travis/virtualenv/python2.7.17/lib/python2.7/site-packages/botocore/client.py", line 316, in _api_call

    return self._make_api_call(operation_name, kwargs)

  File "/home/travis/virtualenv/python2.7.17/lib/python2.7/site-packages/botocore/client.py", line 613, in _make_api_call

    operation_model, request_dict, request_context)

  File "/home/travis/virtualenv/python2.7.17/lib/python2.7/site-packages/botocore/client.py", line 632, in _make_request

    return self._endpoint.make_request(operation_model, request_dict)

  File "/home/travis/virtualenv/python2.7.17/lib/python2.7/site-packages/botocore/endpoint.py", line 102, in make_request

    return self._send_request(request_dict, operation_model)

  File "/home/travis/virtualenv/python2.7.17/lib/python2.7/site-packages/botocore/endpoint.py", line 135, in _send_request

    request, operation_model, context)

  File "/home/travis/virtualenv/python2.7.17/lib/python2.7/site-packages/botocore/endpoint.py", line 167, in _get_response

    request, operation_model)

  File "/home/travis/virtualenv/python2.7.17/lib/python2.7/site-packages/botocore/endpoint.py", line 218, in _do_get_response

    response_dict, operation_model.output_shape)

  File "/home/travis/virtualenv/python2.7.17/lib/python2.7/site-packages/botocore/parsers.py", line 242, in parse

    parsed = self._do_parse(response, shape)

  File "/home/travis/virtualenv/python2.7.17/lib/python2.7/site-packages/botocore/parsers.py", line 508, in _do_parse

    root)

  File "/home/travis/virtualenv/python2.7.17/lib/python2.7/site-packages/botocore/parsers.py", line 515, in _find_result_wrapped_shape

    return mapping[element_name]

KeyError: u'AssumeRoleWithSAMLResult'
```

**Comment by bblommers** [comment]
It looks like the ```AssumeRoleWithSAMLResult``` is not a valid/existing element, based on the error message.

Looking at the offical service JSON, all attributes (Audience, AssumedRoleUser, etc) need to be top-level, instead of inside another XML tag:
https://github.com/boto/botocore/blob/cbd95b0db2c6b03f35d2db79c94c9fa40bd6ad43/botocore/data/sts/2011-06-15/service-2.json#L256

**Comment by microe** [comment]
Ah, I was pulling the raw response, I probably needed to snip the root element. Thanks for looking into it.

**Comment by microe** [comment]
I tried doing that and I ended up breaking the test locally:
```
>       assume_role_response = client.assume_role_with_saml(
            RoleArn=role_input,
            PrincipalArn=principal_role,
            SAMLAssertion=b64encode(saml_assertion.encode("utf-8")).decode("utf-8"),
        )

tests/test_sts/test_sts.py:200: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/moto-dev-env/lib/python3.8/site-packages/botocore/client.py:316: in _api_call
    return self._make_api_call(operation_name, kwargs)
/moto-dev-env/lib/python3.8/site-packages/botocore/client.py:612: in _make_api_call
    http, parsed_response = self._make_request(
/moto-dev-env/lib/python3.8/site-packages/botocore/client.py:632: in _make_request
    return self._endpoint.make_request(operation_model, request_dict)
/moto-dev-env/lib/python3.8/site-packages/botocore/endpoint.py:102: in make_request
    return self._send_request(request_dict, operation_model)
/moto-dev-env/lib/python3.8/site-packages/botocore/endpoint.py:134: in _send_request
    success_response, exception = self._get_response(
/moto-dev-env/lib/python3.8/site-packages/botocore/endpoint.py:166: in _get_response
    success_response, exception = self._do_get_response(
/moto-dev-env/lib/python3.8/site-packages/botocore/endpoint.py:217: in _do_get_response
    parsed_response = parser.parse(
/moto-dev-env/lib/python3.8/site-packages/botocore/parsers.py:242: in parse
    parsed = self._do_parse(response, shape)
/moto-dev-env/lib/python3.8/site-packages/botocore/parsers.py:506: in _do_parse
    start = self._find_result_wrapped_shape(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <botocore.parsers.QueryParser object at 0x7f0370dbe2e0>, element_name = 'AssumeRoleWithSAMLResult'
xml_root_node = <Element '{https://sts.amazonaws.com/doc/2011-06-15/}AssumeRoleWithSAMLResponse' at 0x7f0370de12c0>

    def _find_result_wrapped_shape(self, element_name, xml_root_node):
        mapping = self._build_name_to_xml_node(xml_root_node)
>       return mapping[element_name]
E       KeyError: 'AssumeRoleWithSAMLResult'
```

Is there a way I can run the test the CI system is running? It looks like ``make test-only``. But when I run that make directive, test passes.

**Comment by microe** [comment]
Ah, I see your comment in the PR on how to do server mode. I will give that a try.

**Comment by microe** [comment]
Using your directions from the PR I am able to recreate this failure. I am a little lost because it is failing on a KeyError which means that it can't find a 'AssumeRoleWithSAMLResult' in the response from the moto server.

**Comment by microe** [comment]
Maybe it is some kind of endpoint mangling thing? The endpoint is AssumeRoleWithSAML where the SAML is all caps. Is the server mangling assume_role_with_saml into AssumeRoleWithSaml?

**Comment by microe** [comment]
I am still stuck on this if anyone has any advice. When running in server mode moto does not recognize the API call and then does not send back the AssumeRoleWithSAML response. Am I supposed to do something to connect my sts.STSBackend.assume_role_with_saml method to the server? Is there an easy way to debug the server to see more of what it is doing?

**Comment by bblommers** [comment]
Hi @microe, I think I found the issue.
Moto uses the authorization-header as a hint which service to redirect to. Because the AssumeRoleWithSAML-request is unsigned, Moto didn't know which service to redirect to - so it redirected to the default service, S3.

The default action for S3 is ListBuckets. The error that we we're seeing was STS saying: Hey, I've got a response saying ```<ListAllBuckets>..</ListAllBuckets>```, but I'm expecting the key ```AssumeRoleWithSAMLResponse```

I've pushed a fix directly to your branch - hope you don't mind - that uses the Action to determine which service to use.
